consensus/src/validation.rs

```diff
@@ -0,0 +1,483 @@
+use std::fmt::{Debug, Display, Formatter, Result as FmtResult};
+
+use crate::{
+    alerts::{Alert, ForkingNotification},
+    units::{
+        SignedUnit, UncheckedSignedUnit, Unit, UnitStore, UnitStoreStatus, ValidationError,
+        Validator as UnitValidator, WrappedUnit,
+    },
+    Data, Hasher, MultiKeychain, NodeIndex, NodeSubset, Round,
+};
+
+/// What can go wrong when validating a unit.
+#[derive(Eq, PartialEq)]
+pub enum Error<H: Hasher, D: Data, MK: MultiKeychain> {
+    Invalid(ValidationError<H, D, MK::Signature>),
+    Duplicate(SignedUnit<H, D, MK>),
+    Uncommitted(SignedUnit<H, D, MK>),
+    NewForker(Box<Alert<H, D, MK::Signature>>),
+}
+
+impl<H: Hasher, D: Data, MK: MultiKeychain> Debug for Error<H, D, MK> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        use Error::*;
+        match self {
+            Invalid(e) => write!(f, "Invalid({:?})", e),
+            Duplicate(u) => write!(f, "Duplicate({:?})", u.clone().into_unchecked()),
+            Uncommitted(u) => write!(f, "Uncommitted({:?})", u.clone().into_unchecked()),
+            NewForker(a) => write!(f, "NewForker({:?})", a),
+        }
+    }
+}
+
+impl<H: Hasher, D: Data, MK: MultiKeychain> From<ValidationError<H, D, MK::Signature>>
+    for Error<H, D, MK>
+{
+    fn from(e: ValidationError<H, D, MK::Signature>) -> Self {
+        Error::Invalid(e)
+    }
+}
+
+/// The summary status of the validator.
+pub struct ValidatorStatus {
+    processing_units: UnitStoreStatus,
+    known_forkers: NodeSubset,
+}
+
+impl ValidatorStatus {
+    /// The highest round among the units that are currently processing.
+    pub fn top_round(&self) -> Round {
+        self.processing_units.top_round()
+    }
+}
+
+impl Display for ValidatorStatus {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        write!(
+            f,
+            "processing units: ({}), forkers: {}",
+            self.processing_units, self.known_forkers
+        )
+    }
+}
+
+type ValidatorResult<H, D, MK> = Result<SignedUnit<H, D, MK>, Error<H, D, MK>>;
```

Comment:
wdyt about renaming it to `ValidationResult`?

Reply:
It's not `pub`, so it seemed that it should refer to the `Validator` more than to the whole module, but sure if you think that's better I'll do that.